engine/src/main/java/org/hibernate/validator/internal/metadata/provider/AnnotationMetaDataProvider.java

```diff
@@ -445,36 +438,60 @@ private ConstrainedExecutable findExecutableMetaData(Executable executable) {
 	 * constraint descriptors.
 	 *
 	 * @param constrainable The constrainable to check for constraint annotations.
-	 * @param type The element type the constraint/annotation is placed on.
+	 * @param kind The constraint location kind.
+	 *
+	 * @return A list of constraint descriptors for all constraint specified for the given member.
+	 */
+	private List<ConstraintDescriptorImpl<?>> findConstraints(JavaBeanAnnotatedConstrainable constrainable, ConstraintLocationKind kind) {
+		return findConstraints( constrainable, constrainable, kind );
+	}
+
+	/**
+	 * Finds all constraint annotations defined for the given constrainable and returns them in a list of constraint
+	 * descriptors.
+	 *
+	 * @param constrainable The constrainable element (will be the executable for a method parameter).
+	 * @param annotatedElement The annotated element. Usually the same as the constrainable except in the case of method
+	 * parameters constraints when it is the parameter.
+	 * @param kind The constraint location kind.
 	 *
 	 * @return A list of constraint descriptors for all constraint specified for the given member.
 	 */
-	private List<ConstraintDescriptorImpl<?>> findConstraints(JavaBeanAnnotatedConstrainable constrainable, ConstraintLocationKind type) {
+	private List<ConstraintDescriptorImpl<?>> findConstraints(Constrainable constrainable, JavaBeanAnnotatedElement annotatedElement,
+			ConstraintLocationKind kind) {
 		List<ConstraintDescriptorImpl<?>> metaData = newArrayList();
-		for ( Annotation annotation : constrainable.getDeclaredAnnotations() ) {
-			metaData.addAll( findConstraintAnnotations( constrainable, annotation, type ) );
+		for ( Annotation annotation : annotatedElement.getDeclaredAnnotations() ) {
+			metaData.addAll( findConstraintAnnotations( constrainable, annotation, kind ) );
 		}
 
 		return metaData;
 	}
 
 	/**
-	 * Finds all constraint annotations defined for the given class and returns them in a list of
-	 * constraint descriptors.
+	 * Finds all constraint annotations defined for the given constrainable and returns them in a list of constraint
+	 * descriptors.
 	 *
-	 * @param beanClass The class to check for constraints annotations.
+	 * @param constrainable The constrainable element (will be the executable for a method parameter).
+	 * @param annotations The annotations.
+	 * @param kind The constraint location kind.
 	 *
-	 * @return A list of constraint descriptors for all constraint specified on the given class.
+	 * @return A list of constraint descriptors for all constraint specified for the given member.
 	 */
-	private List<ConstraintDescriptorImpl<?>> findClassLevelConstraints(Class<?> beanClass) {
+	private List<ConstraintDescriptorImpl<?>> findConstraints(Constrainable constrainable, Annotation[] annotations,
+			ConstraintLocationKind kind) {
+		if ( annotations.length == 0 ) {
+			return Collections.emptyList();
+		}
+
 		List<ConstraintDescriptorImpl<?>> metaData = newArrayList();
```

Comment:
Minor style question, mostly out of curiousity: did you actively decide against using streams and map() + collect()? I guess I'd have done that if I were touch the code anyways, but I suppose you have your reasons. It's an honest question, just interested :)

Reply:
The main reason is that using streams usually leads to use lambdas or method references which consumes memory with an additional class loader (Sanne sent us a few emails about this).

Using streams also has its own overhead and, while I'm willing to pay for it when it makes things lots easier, I'm not when a simple for loop does the trick.

In this example https://github.com/hibernate/hibernate-validator/pull/973/files#diff-a0952ad4963831a974f6b5fa78895a6dL762 , not using a stream also helps evacuating the empty case early.

So, my rule of thumb is why not use them when they make things really really nicer but avoid them when they are not a big improvement.

So when I change a code area containing a stream, I consider it carefully and remove it if I don't think it's a clear win.